src/future/shared.rs

```diff
@@ -38,64 +39,76 @@ impl<F> fmt::Debug for Shared<F>
 {
     fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
         fmt.debug_struct("Shared")
-            .field("id", &self.id)
             .field("inner", &self.inner)
             .finish()
     }
 }
 
 struct Inner<F: Future> {
-    next_clone_id: Mutex<u64>,
-    state: Mutex<State<F>>,
+    next_clone_id: AtomicUsize,
+    future: UnsafeCell<Spawn<F>>,
+    result: UnsafeCell<Option<Result<SharedItem<F::Item>, SharedError<F::Error>>>>,
+    unparker: Arc<Unparker>,
 }
 
-impl<F> fmt::Debug for Inner<F>
-    where F: Future + fmt::Debug,
-          F::Item: fmt::Debug,
-          F::Error: fmt::Debug,
-{
-    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
-        fmt.debug_struct("Inner")
-            .field("next_clone_id", &self.next_clone_id)
-            .field("state", &self.state)
-            .finish()
-    }
+struct Unparker {
+    state: AtomicUsize,
+    waiters: Mutex<HashMap<usize, Task>>,
```

Comment:
The previous implementation used `Mutex<HashMap<...>>`... I was trying to do as little work as possible to get `Shared` off of w/ unpark events.